File: Gener.py
import sys

class Automata:
    def __init__(self, estados, alfabeto, transiciones, estado_inicial, estados_finales):
        self.estados = estados
        self.alfabeto = alfabeto
        
        # Validar origen de transiciones
        if not all(origen in estados for (origen, letra) in transiciones.keys()):
            print("Las transiciones deben estar con estados validos")
            sys.exit()
            
        if not all(destino in estados for destino in transiciones.values()):
            print("Las transiciones deben estar con destinos validos")
            sys.exit()
        self.transiciones = transiciones

        if estado_inicial not in estados:
            print("El estado inicial no está en el conjunto")
            sys.exit()

        self.estado_inicial = estado_inicial

        if not all(estado in estados for estado in estados_finales):
            print("Los estados finales no están en el conjunto")
            sys.exit()

        self.estados_finales = estados_finales

    def acepta(self, cadena):
        estado_actual = self.estado_inicial
        for simbolo in cadena:
            if (estado_actual, simbolo) not in self.transiciones:
                print(f"No hay transicion definida para ({estado_actual}, {simbolo})")
                return False
            estado_actual = self.transiciones[(estado_actual, simbolo)]

        return estado_actual in self.estados_finales

def main():
    if len(sys.argv) != 3:
        print("Hace falta uno o mas archivos ")
        return

    caracteres = str.maketrans("", "", "{}\n")
    nombre_archivo = sys.argv[1]
    archivo_cadenas = sys.argv[2]

    try:
        with open(nombre_archivo, "r") as f:
            lineas = f.readlines() 
       
            if len(lineas) != 5:
                print("El archivo debe tener 5 lineas")
                return

            estados = lineas[0].translate(caracteres).split(",")
            alfabeto = lineas[1].translate(caracteres).split(",")
            estados = [e.strip() for e in lineas[0].translate(caracteres).split(",")]
            alfabeto = [a.strip() for a in lineas[1].translate(caracteres).split(",")]

            transiciones_raw = lineas[2].translate(caracteres).split(",")
            transiciones = {}
            for t in transiciones_raw:
                if not t.strip():
                    continue

                izquierda, destino = t.split("=", 1)
                izquierda = izquierda.strip("()") 

                izquierda = izquierda.split(":")

                transiciones[(izquierda[0], izquierda[1])] = destino

            estado_inicial = lineas[3].translate(caracteres).strip()
            estados_finales = lineas[4].translate(caracteres).split(",")
            automata = Automata(estados, alfabeto, transiciones, estado_inicial, estados_finales)
            print("Automata creado correctamente")
            print("Estados:", automata.estados)
            print("Alfabeto:",automata.alfabeto)
            print("Transiciones:", automata.transiciones)
            print("Estado inicial:", automata.estado_inicial)
            print("Estados finales:", automata.estados_finales)
            

            with open(archivo_cadenas, "r") as f:
                cadenas = [linea.strip() for linea in f.readlines() if linea.strip()]
            for cadena in cadenas:
                resultado = automata.acepta(cadena)
                print('ACEPTADA' if resultado else 'RECHAZADA')

    except FileNotFoundError:
        print("Archivo no encontrado.")
    except Exception as e:
        print(f"Ocurrio un error: {e}")

if __name__ == "__main__":
    main()

File: Gener.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ESTADOS 50
#define MAX_ALFABETO 10
#define MAX_TRANSICIONES 200
#define MAX_CADENA 100

typedef struct {
    char origen[10];
    char simbolo[10];
    char destino[10];
} Transicion;


typedef struct {
    char estados[MAX_ESTADOS][10];
    int num_estados;

    char alfabeto[MAX_ALFABETO][10];
    int num_alfabeto;

    Transicion transiciones[MAX_TRANSICIONES];
    int num_transiciones;

    char estado_inicial[10];

    char estados_finales[MAX_ESTADOS][10];
    int num_finales;
} Automata;

int es_final(Automata *a, const char *estado) {
    for (int i = 0; i < a->num_finales; i++) {
        if (strcmp(a->estados_finales[i], estado) == 0)
            return 1;
    }
    return 0;
}


char *transicion(Automata *a, const char *estado, const char *simbolo) {
    for (int i = 0; i < a->num_transiciones; i++) {
        if (strcmp(a->transiciones[i].origen, estado) == 0 &&
            strcmp(a->transiciones[i].simbolo, simbolo) == 0) {
            return a->transiciones[i].destino;
        }
    }
    return NULL;
}


int acepta(Automata *a, const char *cadena) {
    char estado_actual[10];
    strcpy(estado_actual, a->estado_inicial);

    for (int i = 0; i < strlen(cadena); i++) {
        char simbolo[2] = { cadena[i], '\0' };
        char *nuevo_estado = transicion(a, estado_actual, simbolo);

        if (nuevo_estado == NULL) {
            printf("No hay transicion definida para (%s, %s)\n", estado_actual, simbolo);
            return 0;
        }
        strcpy(estado_actual, nuevo_estado);
    }
    return es_final(a, estado_actual);
}


int split_line(char *linea, char tokens[][10], const char *delim) {
    int count = 0;
    char *token = strtok(linea, delim);
    while (token != NULL) {
        strcpy(tokens[count++], token);
        token = strtok(NULL, delim);
    }
    return count;
}

int main(int argc, char *argv[]) {
    if (argc != 3) {
        printf("Uso: %s <archivo_automata> <archivo_cadenas>\n", argv[0]);
        return 1;
    }

    FILE *f = fopen(argv[1], "r");
    if (!f) {
        printf("No se pudo abrir el archivo de automata.\n");
        return 1;
    }

    Automata a;
    a.num_estados = a.num_alfabeto = a.num_transiciones = a.num_finales = 0;

    char linea[1024];


    fgets(linea, sizeof(linea), f);
    linea[strcspn(linea, "\n")] = 0;
    a.num_estados = split_line(linea, a.estados, ",");


    fgets(linea, sizeof(linea), f);
    linea[strcspn(linea, "\n")] = 0;
    a.num_alfabeto = split_line(linea, a.alfabeto, ",");


    fgets(linea, sizeof(linea), f);
    linea[strcspn(linea, "\n")] = 0;

    char partes[MAX_TRANSICIONES][50];
    int num = split_line(linea, partes, ",");
    for (int i = 0; i < num; i++) {
        char izq[20], der[20];
        sscanf(partes[i], "(%[^:]:%[^)])=%s", 
               a.transiciones[a.num_transiciones].origen, 
               a.transiciones[a.num_transiciones].simbolo, 
               a.transiciones[a.num_transiciones].destino);
        a.num_transiciones++;
    }

    fgets(linea, sizeof(linea), f);
    linea[strcspn(linea, "\n")] = 0;
    strcpy(a.estado_inicial, linea);

    fgets(linea, sizeof(linea), f);
    linea[strcspn(linea, "\n")] = 0;
    a.num_finales = split_line(linea, a.estados_finales, ",");

    fclose(f);

    printf("Automata cargado correctamente.\n");

    // Leer archivo de cadenas
    f = fopen(argv[2], "r");
    if (!f) {
        printf("No se pudo abrir el archivo de cadenas.\n");
        return 1;
    }

    while (fgets(linea, sizeof(linea), f)) {
        linea[strcspn(linea, "\n")] = 0;
        if (strlen(linea) == 0) continue;

        int resultado = acepta(&a, linea);
        printf("%s -> %s\n", linea, resultado ? "ACEPTADA" : "RECHAZADA");
    }

    fclose(f);
    return 0;
}
